core/src/main/scala/pureconfig/DerivedReaders.scala

```diff
@@ -198,6 +213,61 @@ trait DerivedReaders1 {
     }
   }
 
+  // We introduce this special type here to keep accumulating the numbers for
+  // building the `WrongSizeList` ConfigReaderFailure. We can't rely on the
+  // `WrongSizeList` type alone since we can have HLists inside HLists and we
+  // don't wan't to keep accumulating for inner HLists.
+  protected[pureconfig] trait HListConfigReader[T <: HList] {
```

Comment:
Visiting the whole `ConfigList` allowed us to potentially fetch more errors. However, I agree that it does complicate the code and the usefulness of it is debatable. I've modified it in fa049e5c0c6b079a2545d498cf1ac59c9449301f.